### PR TITLE
root/array.d: convert data from pointer to array

### PR DESCRIPTION
**Pull request by WalterBright**
Make `Array` more memory safe.

### CODE REVIEW

File: src/dmd/root/array.h [hunk 1]
```diff
@@ -16,12 +16,12 @@ template <typename TYPE>
 struct Array
 {
     d_size_t dim;
-    TYPE *data;
 
   private:
     Array(const Array&);
 
     d_size_t allocdim;
+    TYPE *data;
```
Comment (Geod24) [inline src/dmd/root/array.h]: This should be a `DArray`

**Review by Geod24 — DISMISSED** [review]

File: src/dmd/root/array.h (hunk 1 above)
Comment (WalterBright) [reply src/dmd/root/array.h]: That would break users of it, which I don't have the code for. So I'll let those users make such changes.

**Comment by thewilsonator** [comment]
That field is private, there are no users of it.

**Comment by WalterBright** [comment]
> That field is private, there are no users of it.

When I accidentally had the fields reversed in the .h file, various things seg faulted when running the tests. So _something_ is relying on it as is.

**Comment by thewilsonator** [comment]
Obviously. Those users are LDC and GDC. 

What I meant is that with them in the correct order and being private, then there will be no direct users and it is better to have the header mirror the D source than have some differences.

File: src/dmd/root/array.h (hunk 1 above)
Comment (ibuclaw) [reply src/dmd/root/array.h]: Leaving it as a pointer would break things anyway as Arras type sizes would differ between C++ and D.

**Comment by WalterBright** [comment]
If I change them, I break the LDC/GCD builds. So I'd prefer to let those teams change this.

**Comment by WalterBright** [comment]
Any way to just restart CyberShadow/DAutoTest without rerunning everything else?

**Comment by wilzbach** [comment]
The dub registry is down (https://code.dlang.org), so restarting won't help.

**Comment by Geod24** [comment]
> If I change them, I break the LDC/GCD builds. So I'd prefer to let those teams change this.

I don't think it would scale to leave it up to the GDC/LDC team to find out the changes and adapt the headers.

**Comment by WalterBright** [comment]
If I change it, then the test suite breaks until someone fixes GDC / LDC.

File: src/dmd/root/array.h (hunk 1 above)
Comment (ibuclaw) [reply src/dmd/root/array.h]: It took me a while to work out that you left in `allocdim`, so the sizes do indeed match in both C++ and D, thus nothing breaks.

If you are really that uncomfortable making identical changes to both the C++ and D implementation of Array, then I can fix things up in a follow-up.  As most of Array is extern(C++) anyway, we could even remove the inline implementations in the headers.

**Review by ibuclaw — APPROVED** [review]

**Comment by ibuclaw** [comment]
Rebased against master and pushed.

**Pull request merged**